_layouts/default.html

```diff
@@ -199,4 +199,18 @@ <h2 class="text-delta">Table of contents</h2>
     {% endif %}
   </div>
 </body>
+{% if site.mermaid_enabled != false %}
+  <script>
+    var config = {
+      startOnLoad:true,
+      theme: '{{ site.mermaid.theme }}',
+      flowchart:{
+        useMaxWidth:false,
+        htmlLabels:true
+      }
+    };
```

Comment:
Is there a way we could allow users to create their own custom config? Perhaps with a Jekyll `_includes` that loads into this script tag? I think that was a request from @pdmosses in #825.

Reply:
I'm honestly not very familiar with what is possible here. I could certainly put in an _includes/mermaid_init.html that contains
`<script>
  var config = {
    startOnLoad: true,
    theme: '{{ site.mermaid.theme }}',
    flowchart: {
      useMaxWidth: false,
      htmlLabels: true
    }
  };
  mermaid.initialize(config);
  window.mermaid.init(undefined, document.querySelectorAll('.language-mermaid'));
</script>`
This would allow overriding the entire init process.

I could also move window.mermaid.init(undefined, document.querySelectorAll('.language-mermaid')); into its own script. That would mean mermaid_initialize would include
`<script>
  var config = {
    startOnLoad: true,
    theme: '{{ site.mermaid.theme }}',
    flowchart: {
      useMaxWidth: false,
      htmlLabels: true
    }
  };
  mermaid.initialize(config);
</script>`

I don't know if there is a way to just do the config object which would be ideal.

What would you recommend?

Reply:
@nascosto perhaps the simplest would be to let the user specify all the mermaid config components that can sensibly be varied, setting the rest (with suitable defaults) in your script. For example, based on the config you gave above, a user might specify:

```yaml
mermaid:
  theme: "..."
  flowchart:
    useMaxWidth: true
```

and the script could be something like;

```html
<script>
  var config = {
    startOnLoad: true,
    theme: '{{ site.mermaid.theme | default: ... }}',
    flowchart: {
      useMaxWidth: {{ site.mermaid.flowchart.useMaxWidth | default: false }},
      htmlLabels: {{ site.mermaid.flowchart.htmlLabels | default: true }}
    }
  };
  mermaid.initialize(config);
  window.mermaid.init(undefined, document.querySelectorAll('.language-mermaid'));
</script>
```

If mermaid gets new options in future, it would also be easy to update your script to allow them to be user-specified in a backwards-compatible way.

_Caveats_: I'm not familiar with the mermaid config options, and I haven't tested whether the above would work.

Reply:
I added all the settings that were publicly documented. Give it a look.

Reply:
> I added all the settings that were publicly documented. Give it a look.

Ah! That may have been more than necessary. I don't think it's a good idea for us to create all of these site variables and enforcing defaults; I was thinking either:

- having the `include` just include the config object (you should be able to include this like any old include)
- having the `include` be the entire script tag

From my perspective, all of these site variables can be confusing/hard to maintain, especially keeping track of defaults and library changes (let's say the user changes the version to something else, like an older version that requires a deprecated/removed key. how would that work? would they (and we) have to update the theme version every time mermaid adds something new?).

Is there a specific reason you suggested this approach @pdmosses / @nascosto?

Reply:
@mattxwang  Per the notes down below and the changes made to the PR, is there anything else you'd like done to resolve this?